### ISSUE TITLE
Support for skins that include Angular 1.2?

### ISSUE THREAD

**Issue by janjonas** [2sic/2sxc#382]
We need to support IE8 and thats why we use Angular 1.2 for our skin and our modules.
As far as I've seen your 2sxc module includes Angular 1.3.15. Is there any built-in compatibility mode? In other words: Can I disable your Angular 1.3 version and 2sxc uses our (old) Angular 1.2?

If I insert your App module in our skin, I get the error message:
```
Unknown provider: $translateStaticFilesLoaderProvider <- $translateStaticFilesLoader
```
which seems to be a result of the two different Angular includes.

**Comment by iJungleboy** [comment]
I believe it should work. The thing that seems to be missing is angular-translate, but I'm guessing it's compatible, yet just missing in the downloads. Did you remove our angular-files or not?

**Comment by iJungleboy** [comment]
or what's more likely
1. the first angular loads
2. then translate loads
3. then the second angular loads, and kills the list of modules which the first angular had loaded


...so maybe just use fiddler to check that, and briefly rename our angular-files so they don't get loaded to try what happens...

**Comment by janjonas** [comment]
Hi @iJungleboy,

Thanks for your answer. I've examined the problem in detail and found out that the problem is caused by the version of the angular-translate library that we are using (version 2.4.1 in detail) - your 2sxc module uses version 2.7.2 (included in DesktopModules\ToSIC_SexyContent\dist\lib\i18n\set.min.js).

To reproduce our problem install 2sxc 08.00.01 in a new DNN 7.4.2 environment and edit the Default.aspx and add the following tag before the closing </body>:

<script src="https://cdnjs.cloudflare.com/ajax/libs/angular-translate/2.4.1/angular-translate.js" type="text/javascript"></script>

Of course, modifying the Default.aspx file is not best practice but it is an easy way to demonstrate the conflict.

Wouldn't it be nice if your "app selector" module does not include any 3rd party libraries (like Angular) but just uses DNN default libraries (like jQuery) or just "plain" JavaScipt? I know, you do not like jQuery, but including new libraries could always result in conflicts with other parts of the website (modules, skins, ...). In your edit (inline popups), 3rd party libraries (like Angular) aren't a problem because they use iframes and run in there own "sandbox".
IMHO compatibility (to other modules and skins) is much more important than using Angular for rendering the app selector module :)

**Comment by janjonas** [comment]
FYI: Even including the same version (2.7.2) in Default.aspx just before </body> breaks 2SXC:

```
<script src="https://cdnjs.cloudflare.com/ajax/libs/angular-translate/2.7.2/angular-translate.js" type="text/javascript"></script>  
```

**Comment by iJungleboy** [comment]
@janjonas 
I basically agree that we shouldn't include angular in the page. Problem is that when we started using it, there was no good convention for handling it. I believe now the client dependency has something, but I don't know if it's reliable and if angular is a "default" package or if we would have to supply it. 

If you want to look into that we would love to include the change - I just don't have time right now to review it and since 2sxc 8 runs in a completely separate window we don't have these problems there (no dnn page around it). So the problem is only left to the little bit in-page, and that would be great if we could do it in a nicer way. 

In general: are you sure it's a version problem, or could it also just be a sequence problem? like 
1. angular loads
2. registers translate
3. angular loads again - loses previous registrations
4. ...?

**Comment by janjonas** [comment]
Hi @iJungleboy,
since the problem also exists when including the same version of angular-translate it has to be caused by the duplicate import or by the order of the imports. I'm not sure if client dependency really helps. IMHO your in-page content should not rely on any external dependencies to guarantee a maximum of compatibility to any other skin, modules, library, script or whatever. It should just work :).
In the end that would mean either put an iframe around it our build it with jQuery or plain JavaScript.
If I find the time, I will deep dive into your code and see what I can do.
Thanks
  Jan

**Comment by iJungleboy** [comment]
@janjonas I believe a change to client dependency would fix it, because that would ensure that it's only loaded once - assuming it lies in the same folder. So the key would be to create client-dependency packages and then tell both the other app as well as 2sxc to use these.

**Comment by janjonas** [comment]
Hi @iJungleboy: Sorry for my late reply. Are you talking about JavaScript libraries that can be installed in DNN (stored in ~/Resources/Libraries\...)? I'm not quite sure if this really solves the problem since there is no guarantee that your requested angular version and angular version that are needed by other skins and/or modules are compatible. Also the include order of scripts could cause various problems.

You are using iframes everywhere to make sure 2sxc runs in his own world which is great. Why not use this concept for the "app selector"? IMHO this is the only place where client side resources could cause conflicts. 
Another approach would be to use jQuery for the "app selector". IMHO, there are just two simple dropdowns that reload the complete page after an entry was selected. So there is no need complex functionality that requires AngularJS.

**Comment by iJungleboy** [comment]
It's actually more complex that that. The app selector has multi-language and will soon get more visual - so Angular is the way to go. 
Also, the toolbar actually talks to the app-selector because the toolbar isn't angular but needs some things only the app-selector knows. 

So I believe the correct way is to

1. Create a scripts-library package containing angular, which 2sxc would distribute in the installation to ensure that at least 1 version exists - I'm not sure but I believe that this package may already exist...
2. then change the way the scripts are referenced to ensure it tries to take those

**Comment by janjonas** [comment]
I'm not sure if this really helps: What about AngularJS version conflicts and/or conflicts with the order the scripts are included?

Isn't it possible to put your app selector into an iframe if it requires AngularJS?

**Comment by iJungleboy** [comment]
Script orders are +/- normed - that shouldn't be an issue. There are number-sets for various kinds of scripts, and if different systems tell the dependency resolver what they want, it's in charge of fixing it. 

IFraming may work, probably a good idea. But it would be a fairly large refactoring (because the toolbar and the menu also "talk" to that application) and quite a bit of testing to be really sure it works, and it may have side-effects with future wishes like live-preview. But in general I believe it would be a very nice idea - also because it "protects" my application from any side-effects. My guess though is that it's 1-2 days of work to get it all to work.

**Comment by iJungleboy** [comment]
Will be implemented in 8.4, entire ui is now in an iframe
